java/src/Ice/src/main/java/com/zeroc/IceInternal/RequestHandlerCache.java

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) ZeroC, Inc.
+
+package com.zeroc.IceInternal;
+
+import com.zeroc.Ice.Connection;
+
+public class RequestHandlerCache {
+  RequestHandler getRequestHandler() {
+    if (_cacheConnection) {
+      synchronized (this) {
+        if (_cachedRequestHandler != null) {
+          return _cachedRequestHandler;
+        }
+      }
+    }
+
+    RequestHandler handler = _reference.getRequestHandler();
+    if (_cacheConnection) {
+      synchronized (this) {
+        if (_cachedRequestHandler == null) {
+          _cachedRequestHandler = handler;
+        }
+        // else ignore handler: another thread cached a handler before we could.
+        return _cachedRequestHandler;
+      }
+    } else {
+      return handler;
+    }
+  }
+
+  public Connection getCachedConnection() {
+    if (_cacheConnection) {
+      RequestHandler handler;
+      synchronized (this) {
+        handler = _cachedRequestHandler;
+      }
+      if (handler != null) {
+        return handler.getConnection();
+      }
+    }
+    return null;
+  }
+
+  void clearCachedRequestHandler(RequestHandler handler) {
+    if (_cacheConnection) {
+      synchronized (this) {
+        if (handler == _cachedRequestHandler) {
+          _cachedRequestHandler = null;
+        }
+      }
+    }
+  }
+
+  public RequestHandlerCache(Reference reference) {
+    _reference = reference;
+    _cacheConnection = reference.getCacheConnection();
+  }
+
+  private final Reference _reference;
```

Comment:
Move to the top?

Reply:
In this PR I was just trying to be consistent with our current style.
Are we going to start putting fields at the top going forward?

I'm all for that, since I think that's the dominant convention in Java. Just want to make sure we're on the same page.